File: deploy/cloudflare/src/routes/lemma.ts
/**
 * Lemma Intelligence API Routes
 *
 * Provides dictionary lookup and corpus statistics for Greek and Latin lemmas.
 * Supports:
 * - LSJ (Greek-English) and Lewis & Short (Latin-English) dictionary lookup
 * - Corpus statistics: occurrence counts, author/work distribution
 * - Fuzzy search with trigram matching
 * - Related lemmas by co-occurrence
 * - Knowledge graph connections
 */

import { Hono } from 'hono';
import { Env } from '../types';
import { getLogger } from '../utils/logger';

const logger = getLogger('LemmaRoutes');

export const lemmaRoutes = new Hono<{ Bindings: Env }>();

/**
 * Normalize Greek text for search (lowercase, no diacritics)
 */
function normalizeGreek(text: string): string {
  return text
    .toLowerCase()
    .normalize('NFD')
    .replace(/[\u0300-\u036f]/g, '')  // Remove diacritics
    .normalize('NFC');
}

/**
 * Helper to make Supabase REST API requests for public schema tables
 */
async function supabaseQuery(
  env: Env,
  table: string,
  params: {
    select?: string;
    filters?: string;
    limit?: number;
    order?: string;
  }
): Promise<any[]> {
  const baseUrl = `${env.SUPABASE_URL}/rest/v1`;
  const urlParams = new URLSearchParams();

  if (params.select) {
    urlParams.append('select', params.select);
  }

  if (params.limit) {
    urlParams.append('limit', params.limit.toString());
  }

  if (params.order) {
    urlParams.append('order', params.order);
  }

  let url = `${baseUrl}/${table}?${urlParams.toString()}`;

  // Append raw filters directly to URL
  if (params.filters) {
    url += `&${params.filters}`;
  }

  const headers: Record<string, string> = {
    'apikey': env.SUPABASE_KEY,
    'Authorization': `Bearer ${env.SUPABASE_KEY}`,
    'Content-Type': 'application/json',
  };

  const response = await fetch(url, { headers });

  if (!response.ok) {
    const errorText = await response.text();
    throw new Error(`Supabase query failed: ${response.status} - ${errorText}`);
  }

  return await response.json();
}

/**
 * Helper to call Supabase RPC functions (for accessing free_will schema tables)
 */
async function supabaseRpc(
  env: Env,
  functionName: string,
  params: Record<string, any>
): Promise<any[]> {
  const url = `${env.SUPABASE_URL}/rest/v1/rpc/${functionName}`;

  const headers: Record<string, string> = {
    'apikey': env.SUPABASE_KEY,
    'Authorization': `Bearer ${env.SUPABASE_KEY}`,
    'Content-Type': 'application/json',
  };

  const response = await fetch(url, {
    method: 'POST',
    headers,
    body: JSON.stringify(params),
  });

  if (!response.ok) {
    const errorText = await response.text();
    throw new Error(`Supabase RPC failed: ${response.status} - ${errorText}`);
  }

  return await response.json();
}

/**
 * GET /dictionary/:lemma
 * Look up a lemma in the dictionary (LSJ for Greek, Lewis & Short for Latin)
 * Uses RPC functions to access free_will schema tables
 */
lemmaRoutes.get('/dictionary/:lemma', async (c) => {
  try {
    const lemma = decodeURIComponent(c.req.param('lemma'));
    const language = c.req.query('language') || 'grc';

    if (language === 'grc') {
      // Greek: search LSJ dictionary via RPC
      const normalized = normalizeGreek(lemma);

      const result = await supabaseRpc(c.env, 'get_dictionary_lsj', {
        p_lemma: lemma,
        p_normalized: normalized,
        p_latin: lemma.toLowerCase(),
      });

      if (result && result.length > 0) {
        const row = result[0];
        return c.json({
          found: true,
          language: 'grc',
          dictionary: 'LSJ',
          lemma: row.lemma,
          lemma_latin: row.lemma_latin,
          definition: row.definition,
          short_def: row.short_def,
          forms: row.forms || [],
          greek_forms: row.greek_forms || [],
          external_links: {
            logeion: `https://logeion.uchicago.edu/${lemma}`,
            perseus: `https://www.perseus.tufts.edu/hopper/morph?l=${lemma}&la=greek`,
            bailly: `https://bailly.app/recherche?q=${lemma}`,
          },
        });
      }
    } else {
      // Latin: search Lewis & Short dictionary via RPC
      const result = await supabaseRpc(c.env, 'get_dictionary_lewis_short', {
        p_lemma: lemma,
        p_normalized: lemma.toLowerCase(),
      });

      if (result && result.length > 0) {
        const row = result[0];
        return c.json({
          found: true,
          language: 'lat',
          dictionary: 'Lewis & Short',
          lemma: row.lemma,
          definition: row.definition,
          short_def: row.short_def,
          entry_key: row.entry_key,
          external_links: {
            logeion: `https://logeion.uchicago.edu/${lemma}`,
            perseus: `https://www.perseus.tufts.edu/hopper/morph?l=${lemma}&la=latin`,
          },
        });
      }
    }

    // Not found
    return c.json({
      found: false,
      language,
      lemma,
      message: `No dictionary entry found for '${lemma}'`,
      external_links: {
        logeion: `https://logeion.uchicago.edu/${lemma}`,
      },
    });
  } catch (error) {
    logger.error('Dictionary lookup error', error);
    return c.json({ error: 'Dictionary lookup failed', details: error instanceof Error ? error.message : 'Unknown error' }, 500);
  }
});

/**
 * GET /dictionary/search/:query
 * Search for dictionary entries by prefix or fuzzy match
 * Uses RPC functions to access free_will schema tables
 */
lemmaRoutes.get('/dictionary/search/:query', async (c) => {
  try {
    const query = decodeURIComponent(c.req.param('query'));
    const language = c.req.query('language') || 'grc';
    const limit = parseInt(c.req.query('limit') || '10', 10);
    const fuzzy = c.req.query('fuzzy') === 'true';

    let results: any[] = [];

    if (language === 'grc') {
      // Greek: search LSJ via RPC (prefix search)
      results = await supabaseRpc(c.env, 'search_dictionary_lsj', {
        p_query: query,
        p_limit: limit,
      });
    } else {
      // Latin: search Lewis & Short via RPC
      results = await supabaseRpc(c.env, 'search_dictionary_lewis_short', {
        p_query: query,
        p_limit: limit,
      });
    }

    return c.json({
      query,
      language,
      fuzzy,
      results: results || [],
      count: results?.length || 0,
    });
  } catch (error) {
    logger.error('Dictionary search error', error);
    return c.json({ error: 'Dictionary search failed', details: error instanceof Error ? error.message : 'Unknown error' }, 500);
  }
});

/**
 * GET /stats/:lemma
 * Get corpus statistics for a lemma
 *
 * Note: This endpoint requires complex JSONB queries which are not easily done
 * via Supabase REST API. Returns simplified stats or uses fallback.
 */
lemmaRoutes.get('/stats/:lemma', async (c) => {
  try {
    const lemma = decodeURIComponent(c.req.param('lemma'));
    const language = c.req.query('language') || 'grc';

    // Return a simplified response since complex JSONB queries
    // are not easily done via Supabase REST API without RPC functions
    return c.json({
      lemma,
      language,
      total_occurrences: 0,
      passage_count: 0,
      by_author: [],
      by_work: [],
      by_period: [],
      note: 'Full stats require FastAPI backend. Use /api/lemma/stats/:lemma endpoint on localhost:8000 for complete statistics.',
    });
  } catch (error) {
    logger.error('Lemma stats error', error);
    return c.json({ error: 'Lemma stats lookup failed' }, 500);
  }
});

/**
 * GET /related/:lemma
 * Find lemmas that frequently co-occur with the given lemma
 *
 * Note: This endpoint requires complex JSONB queries which are not easily done
 * via Supabase REST API. Returns simplified response.
 */
lemmaRoutes.get('/related/:lemma', async (c) => {
  try {
    const lemma = decodeURIComponent(c.req.param('lemma'));
    const language = c.req.query('language') || 'grc';

    // Return a simplified response
    return c.json({
      lemma,
      language,
      related: [],
      note: 'Full related lemmas require FastAPI backend. Use /api/lemma/related/:lemma endpoint on localhost:8000 for complete data.',
    });
  } catch (error) {
    logger.error('Related lemmas error', error);
    return c.json({ error: 'Related lemmas lookup failed' }, 500);
  }
});

/**
 * GET /kg-connections/:lemma
 * Find knowledge graph nodes whose descriptions contain this lemma
 */
lemmaRoutes.get('/kg-connections/:lemma', async (c) => {
  try {
    const lemma = decodeURIComponent(c.req.param('lemma'));
    const language = c.req.query('language') || 'grc';

    // Search KG nodes by description containing the lemma (using ILIKE)
    // Note: For KG nodes we use public schema (default)
    // Columns: node_id (PK), label, type, description
    const nodes = await supabaseQuery(c.env, 'kg_nodes', {
      select: 'node_id,label,type,description',
      limit: 10,
      filters: `or=(description.ilike.*${encodeURIComponent(lemma)}*,label.ilike.*${encodeURIComponent(lemma)}*)`,
      order: 'label.asc',
    });

    return c.json({
      lemma,
      language,
      kg_nodes: (nodes || []).map((r: any) => ({
        node_id: r.node_id,
        label: r.label,
        type: r.type,
        description: r.description?.length > 200
          ? r.description.slice(0, 200) + '...'
          : r.description,
      })),
    });
  } catch (error) {
    logger.error('KG connections error', error);
    return c.json({ error: 'KG connections lookup failed', details: error instanceof Error ? error.message : 'Unknown error' }, 500);
  }
});

File: backend/routes/lemma.py
"""
Lemma intelligence routes — dictionary lookup, corpus stats, co-occurrences, KG connections.

Queries the existing passages/oga_tokens tables and the knowledge graph.
No new services needed — all SQL-based.
"""

import logging
from typing import Annotated, Any
from urllib.parse import quote

from fastapi import APIRouter, Depends, Query

from backend.dependencies import get_db
from eleutheria_database.services.db import DatabaseService

logger = logging.getLogger(__name__)

router = APIRouter(tags=["lemma"])


@router.get("/dictionary/{lemma}")
async def get_lemma_dictionary(
    lemma: str,
    db: Annotated[DatabaseService, Depends(get_db)],
    language: str = Query("grc", description="Language: grc or lat"),
) -> dict[str, Any]:
    """
    Look up a lemma in the corpus (mimics LSJ/Lewis & Short dictionary lookup).

    Returns definition-like data from the corpus — occurrence counts,
    common forms, and external links to Logeion and Perseus.
    """
    # Find the lemma in OGA tokens
    row = await db.fetchrow(
        """
        SELECT
            t.lemma, t.pos,
            w.language,
            COUNT(*) as occurrences,
            COUNT(DISTINCT t.work_id) as work_count
        FROM free_will.oga_tokens t
        JOIN free_will.ancient_works w ON t.work_id = w.work_id
        WHERE t.lemma = $1 AND w.language = $2
        GROUP BY t.lemma, t.pos, w.language
        LIMIT 1
        """,
        lemma, language,
    )

    if not row:
        encoded = quote(lemma, safe="")
        return {
            "found": False,
            "language": language,
            "lemma": lemma,
            "message": "Lemma not found in corpus",
            "external_links": {
                "logeion": f"https://logeion.uchicago.edu/{encoded}",
            },
        }

    # Get surface forms
    forms_rows = await db.fetch(
        """
        SELECT DISTINCT surface_form
        FROM free_will.oga_tokens
        WHERE lemma = $1
        LIMIT 20
        """,
        lemma,
    )
    forms = [r["surface_form"] for r in forms_rows]

    encoded = quote(lemma, safe="")
    dict_name = "LSJ" if language == "grc" else "Lewis & Short"

    return {
        "found": True,
        "language": language,
        "dictionary": dict_name,
        "lemma": lemma,
        "lemma_latin": _greek_to_latin(lemma) if language == "grc" else lemma,
        "short_def": f"{row['occurrences']} occurrences across {row['work_count']} works",
        "forms": forms,
        "greek_forms": forms if language == "grc" else [],
        "external_links": {
            "logeion": f"https://logeion.uchicago.edu/{encoded}",
            "perseus": f"https://www.perseus.tufts.edu/hopper/morph?l={encoded}&la={'greek' if language == 'grc' else 'latin'}",
        },
    }


@router.get("/dictionary/search/{query}")
async def search_lemma_dictionary(
    query: str,
    db: Annotated[DatabaseService, Depends(get_db)],
    language: str = Query("grc"),
    limit: int = Query(20, ge=1, le=100),
    fuzzy: bool = Query(False),
) -> dict[str, Any]:
    """Search dictionary entries by prefix or fuzzy match."""
    if fuzzy:
        pattern_sql = "t.lemma ILIKE '%' || $1 || '%'"
    else:
        pattern_sql = "t.lemma ILIKE $1 || '%'"

    rows = await db.fetch(
        f"""
        SELECT
            t.lemma,
            t.pos,
            COUNT(*) as count
        FROM free_will.oga_tokens t
        JOIN free_will.ancient_works w ON t.work_id = w.work_id
        WHERE {pattern_sql}
          AND w.language = $2
          AND t.lemma IS NOT NULL
        GROUP BY t.lemma, t.pos
        ORDER BY COUNT(*) DESC
        LIMIT $3
        """,
        query, language, limit,
    )

    results = []
    for r in rows:
        results.append({
            "lemma": r["lemma"],
            "lemma_latin": _greek_to_latin(r["lemma"]) if language == "grc" else r["lemma"],
            "short_def": f"{r['count']} occurrences",
        })

    return {
        "query": query,
        "language": language,
        "fuzzy": fuzzy,
        "results": results,
        "count": len(results),
    }


@router.get("/stats/{lemma}")
async def get_lemma_stats(
    lemma: str,
    db: Annotated[DatabaseService, Depends(get_db)],
    language: str = Query("grc"),
) -> dict[str, Any]:
    """Get corpus statistics for a lemma — occurrences, distribution by author/work/period."""
    # Total occurrences
    total = await db.fetchval(
        """
        SELECT COUNT(*) FROM free_will.oga_tokens t
        JOIN free_will.ancient_works w ON t.work_id = w.work_id
        WHERE t.lemma = $1 AND w.language = $2
        """,
        lemma, language,
    )

    # Passage count (unique works)
    passage_count = await db.fetchval(
        """
        SELECT COUNT(DISTINCT t.work_id)
        FROM free_will.oga_tokens t
        JOIN free_will.ancient_works w ON t.work_id = w.work_id
        WHERE t.lemma = $1 AND w.language = $2
        """,
        lemma, language,
    )

    # By author
    by_author = await db.fetch(
        """
        SELECT w.author, COUNT(DISTINCT t.work_id) as passages
        FROM free_will.oga_tokens t
        JOIN free_will.ancient_works w ON t.work_id = w.work_id
        WHERE t.lemma = $1 AND w.language = $2
        GROUP BY w.author ORDER BY passages DESC LIMIT 20
        """,
        lemma, language,
    )

    # By work
    by_work = await db.fetch(
        """
        SELECT w.author, w.title, COUNT(*) as passages
        FROM free_will.oga_tokens t
        JOIN free_will.ancient_works w ON t.work_id = w.work_id
        WHERE t.lemma = $1 AND w.language = $2
        GROUP BY w.author, w.title ORDER BY passages DESC LIMIT 20
        """,
        lemma, language,
    )

    # By period
    by_period = await db.fetch(
        """
        SELECT w.period, COUNT(DISTINCT t.work_id) as passages
        FROM free_will.oga_tokens t
        JOIN free_will.ancient_works w ON t.work_id = w.work_id
        WHERE t.lemma = $1 AND w.language = $2 AND w.period IS NOT NULL
        GROUP BY w.period ORDER BY passages DESC
        """,
        lemma, language,
    )

    return {
        "lemma": lemma,
        "language": language,
        "total_occurrences": int(total or 0),
        "passage_count": int(passage_count or 0),
        "by_author": [dict(r) for r in by_author],
        "by_work": [dict(r) for r in by_work],
        "by_period": [dict(r) for r in by_period],
    }


@router.get("/related/{lemma}")
async def get_related_lemmas(
    lemma: str,
    db: Annotated[DatabaseService, Depends(get_db)],
    language: str = Query("grc"),
    limit: int = Query(20, ge=1, le=100),
) -> dict[str, Any]:
    """Get lemmas that frequently co-occur with the given lemma (same work)."""
    rows = await db.fetch(
        """
        SELECT t2.lemma, t2.pos, COUNT(*) as cooccurrences
        FROM free_will.oga_tokens t1
        JOIN free_will.oga_tokens t2
            ON t1.work_id = t2.work_id
            AND t2.lemma != $1
            AND t2.lemma IS NOT NULL
        JOIN free_will.ancient_works w ON t1.work_id = w.work_id
        WHERE t1.lemma = $1 AND w.language = $2
        GROUP BY t2.lemma, t2.pos
        ORDER BY cooccurrences DESC
        LIMIT $3
        """,
        lemma, language, limit,
    )

    return {
        "lemma": lemma,
        "language": language,
        "related": [dict(r) for r in rows],
    }


@router.get("/kg-connections/{lemma}")
async def get_lemma_kg_connections(
    lemma: str,
    db: Annotated[DatabaseService, Depends(get_db)],
    language: str = Query("grc"),
) -> dict[str, Any]:
    """Find KG nodes related to a lemma via passage citations."""
    rows = await db.fetch(
        """
        SELECT DISTINCT
            n.node_id, n.label, n.type, n.description
        FROM free_will.oga_tokens t
        JOIN free_will.passages p ON t.work_id = p.work_id
        JOIN free_will.passage_citations pc ON p.passage_id = pc.passage_id
        JOIN free_will.kg_nodes n ON pc.kg_node_id = n.node_id
        JOIN free_will.ancient_works w ON t.work_id = w.work_id
        WHERE t.lemma = $1 AND w.language = $2
        LIMIT 20
        """,
        lemma, language,
    )

    return {
        "lemma": lemma,
        "language": language,
        "kg_nodes": [dict(r) for r in rows],
    }


# ---------- Helpers ----------

_GREEK_TO_LATIN: dict[str, str] = {
    "\u03b1": "a", "\u03b2": "b", "\u03b3": "g", "\u03b4": "d",
    "\u03b5": "e", "\u03b6": "z", "\u03b7": "h", "\u03b8": "q",
    "\u03b9": "i", "\u03ba": "k", "\u03bb": "l", "\u03bc": "m",
    "\u03bd": "n", "\u03be": "x", "\u03bf": "o", "\u03c0": "p",
    "\u03c1": "r", "\u03c3": "s", "\u03c4": "t", "\u03c5": "u",
    "\u03c6": "f", "\u03c7": "c", "\u03c8": "y", "\u03c9": "w",
    "\u03c2": "s",  # final sigma
}


def _greek_to_latin(text: str) -> str:
    return "".join(_GREEK_TO_LATIN.get(c, c) for c in text.lower())
